core/src/test/java/nu/yona/server/subscriptions/service/migration/MoveVpnPasswordToDeviceTest.java

```diff
@@ -0,0 +1,155 @@
+				@ComponentScan.Filter(pattern = "nu.yona.server.device.service.DeviceService", type = FilterType.REGEX),
+				@ComponentScan.Filter(pattern = "nu.yona.server.Translator", type = FilterType.REGEX) })
+class MoveVpnPasswordToDeviceTestConfiguration extends UserRepositoriesConfiguration
+{
+	@Bean(name = "messageSource")
+	public ReloadableResourceBundleMessageSource messageSource()
+	{
+		return new CoreConfiguration().messageSource();
+	}
+
+	@Bean
+	UserDeviceRepository getMockUserDeviceRepository()
+	{
+		return new UserDeviceRepositoryMock();
+	}
+
+	@Bean
+	DeviceAnonymizedRepository getDeviceAnonymizedRepository()
+	{
+		return new DeviceAnonymizedRepositoryMock();
+	}
+
+	@Bean
+	BuddyAnonymizedRepository getBuddyAnonymizedRepository()
+	{
+		return new BuddyAnonymizedRepositoryMock();
+	}
+}
+
+@RunWith(SpringJUnit4ClassRunner.class)
+@ContextConfiguration(classes = { MoveVpnPasswordToDeviceTestConfiguration.class })
+public class MoveVpnPasswordToDeviceTest extends BaseSpringIntegrationTest
+{
+	private static final String PASSWORD = "password";
+	private static final Field vpnPasswordField = JUnitUtil.getAccessibleField(UserPrivate.class, "vpnPassword");
+	private static final Field userPrivateField = JUnitUtil.getAccessibleField(User.class, "userPrivate");
+	private User richard;
+
+	@Autowired
+	private DeviceAnonymizedRepository deviceAnonymizedRepository;
+
+	@Autowired
+	private MigrationStep migrationStep;
+
+	@Rule
+	public MethodRule cryptoSession = new CryptoSessionRule(PASSWORD);
+
+	@Before
+	public void setUpPerTest() throws Exception
+	{
+		try (CryptoSession cryptoSession = CryptoSession.start(PASSWORD))
+		{
+			richard = JUnitUtil.createRichard();
+		}
+	}
+
+	@Override
+	protected Map<Class<?>, Repository<?, ?>> getRepositories()
+	{
+		Map<Class<?>, Repository<?, ?>> repositoriesMap = new HashMap<>();
+		repositoriesMap.put(DeviceAnonymized.class, deviceAnonymizedRepository);
+		return repositoriesMap;
+	}
+
+	@Test
+	public void upgrade_movePassword_passwordMoved() throws IllegalArgumentException, IllegalAccessException
+	{
+		String vpnPassword = "upgrade_movePassword_passwordMoved";
+		vpnPasswordField.set(userPrivateField.get(richard), vpnPassword);
+
+		UserDevice device = createDevice();
+		richard.addDevice(device);
+
+		migrationStep.upgrade(richard);
+		assertThat(device.getVpnPassword(), equalTo(vpnPassword));
+		assertThat(richard.takeVpnPassword(), equalTo(Optional.empty()));
+	}
+
+	@Test
+	public void upgrade_movePasswordTwice_idemPotent() throws IllegalArgumentException, IllegalAccessException
+	{
+		String vpnPassword = "upgrade_movePasswordTwice_idemPotent";
+		vpnPasswordField.set(userPrivateField.get(richard), vpnPassword);
+
+		UserDevice device = createDevice();
+		richard.addDevice(device);
+
+		migrationStep.upgrade(richard);
+		assertThat(device.getVpnPassword(), equalTo(vpnPassword));
```

Comment:
I would prefer to have a simple act:
```
migrationStep.upgrade(richard);
migrationStep.upgrade(richard);
```
The first two asserts are superfluous in this test. Could we also use the convention to separate arrange/act/assert by newlines?

Reply:
I regularly add assertions to verify the test setup was done right. That's why I added the assertions after the first upgrade. But I'm fine removing them.